src/detect-smb-cmd.c

```diff
@@ -0,0 +1,117 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#include "suricata-common.h"
+
+#include "detect.h"
+#include "detect-parse.h"
+
+#include "detect-engine.h"
+#include "detect-engine-mpm.h"
+#include "detect-engine-state.h"
+#include "detect-engine-prefilter.h"
+#include "detect-engine-content-inspection.h"
+
+#include "detect-smb-cmd.h"
+#include "rust.h"
+
+static int g_smb_cmd_list_id = 0;
+
+static void DetectSmbCmdFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCEnter();
+
+    SCLogDebug("smb_cmd: DetectSmbCmdFree");
+
+    if (ptr != NULL) {
+        rs_smb_cmd_free(ptr);
+    }
+    SCReturn;
+}
+
+static int DetectSmbCmdSetup(DetectEngineCtx *de_ctx, Signature *s, const char *arg)
+{
+    SCLogDebug("smb_cmd: DetectSmbCmdSetup");
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_SMB) < 0)
+        return -1;
+
+    if (arg == NULL) {
+        SCLogError(SC_ERR_INVALID_SIGNATURE, "Error parsing smb_cmd option in "
+                                             "signature, it needs a value");
+        return -1;
+    }
+
+    void *dod = rs_smb_cmd_parse(arg);
+    if (dod == NULL) {
+        SCLogError(SC_ERR_INVALID_SIGNATURE, "Error parsing smb_cmd option in "
+                                             "signature");
+        return -1;
+    }
+
+    SigMatch *sm = SigMatchAlloc();
+    if (sm == NULL) {
+        DetectSmbCmdFree(de_ctx, dod);
+        return -1;
+    }
+
+    sm->type = DETECT_SMB_CMD;
+    sm->ctx = (void *)dod;
+
+    SigMatchAppendSMToList(s, sm, g_smb_cmd_list_id);
+    return 0;
+}
+
+static int DetectSmbCmdMatchRust(DetectEngineThreadCtx *det_ctx, Flow *f, uint8_t flags,
+        void *state, void *txv, const Signature *s, const SigMatchCtx *m)
+{
+    SCEnter();
+
+    SCLogDebug("smb_cmd: DetectSmbCmdMatchRust");
+
+    if (rs_smb_cmd_match(txv, (void *)m) != 1)
+        SCReturnInt(0);
+
+    SCReturnInt(1);
+}
+
+static int DetectEngineInspectSmbCmd(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const struct DetectEngineAppInspectionEngine_ *engine, const Signature *s, Flow *f,
+        uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(
+            de_ctx, det_ctx, s, engine->smd, f, flags, alstate, txv, tx_id);
+}
+
+void DetectSmbCmdRegister(void)
+{
+    sigmatch_table[DETECT_SMB_CMD].name = "smb.cmd";
+    sigmatch_table[DETECT_SMB_CMD].alias = "smb_cmd";
+    sigmatch_table[DETECT_SMB_CMD].desc = "Match SMB message type";
+    sigmatch_table[DETECT_SMB_CMD].Setup = DetectSmbCmdSetup;
+    sigmatch_table[DETECT_SMB_CMD].Match = NULL;
+    sigmatch_table[DETECT_SMB_CMD].AppLayerTxMatch = DetectSmbCmdMatchRust;
+    sigmatch_table[DETECT_SMB_CMD].Free = DetectSmbCmdFree;
+
+    DetectAppLayerInspectEngineRegister2(
```

Comment:
Same as for smb.version and discussed already https://github.com/OISF/suricata/pull/7157#discussion_r835428634

`http2.frametype` does not need this and I do not see the difference...

Reply:
I don't know the difference neither but when these are removed, the tests fail the following way:
```
===> smb-smb_cmd: Sub test #1: FAIL : expected 1 matches; got 11 for filter {'count': 1, 'match': {'event_type': 'alert', 'alert.signature_id': 1}}
===> smb-smb_cmd: Sub test #2: FAIL : expected 2 matches; got 11 for filter {'count': 2, 'match': {'event_type': 'alert', 'alert.signature_id': 2}}
===> smb-smb_cmd: Sub test #3: FAIL : expected 2 matches; got 11 for filter {'count': 2, 'match': {'event_type': 'alert', 'alert.signature_id': 3}}
===> smb-smb_cmd: Sub test #4: FAIL : expected 5 matches; got 11 for filter {'count': 5, 'match': {'event_type': 'alert', 'alert.signature_id': 4}}
```